### PR TITLE
LB-866 (II): Use user_id column to fetch/delete listens

### PR DESCRIPTION
**Pull request by amCap1712**
#1574 added user ids to the listen table. This PR migrates the code to fetch and delete listens to use user ids instead. 

- Listen counts also have a user name column so those can't use user ids yet. #1700 will overhaul listen counts anyways so the intent is to make the user id switch for counts there. At a few places, both user ids and user names are needed currently for this reason. All references to user names will go away in the long run.

- I haven't updated the dumps yet to use usernames from "user" table instead of the listen table because currently we haven't removed the user name column, will do that shift in a separate PR.

- When deploying this, the cache keys for user timestamps will change so we should need to regenerate those in advance to avoid slowdowns in prod. My suggestion is stop ts writer prod, run a script from web beta to generate new redis keys, update ts writer prod to use new keys, bring up new ts writer and web.

### CODE REVIEW

File: admin/timescale/updates/2022-01-13-add-user-id-indexes.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+-- CREATE INDEX will lock the whole table, which means that it'll be unavailable while the index
+-- is building. Timescale doesn't support CREATE INDEX CONCURRENTLY. The best we can do is this
+-- extension which opens a transaction and creates the index one chunk at a time.
+-- This lets inserts and selects work on the table while it's running
+-- This operation doesn't support being run inside a transaction.
+CREATE INDEX listened_at_user_id_ndx_listen ON listen (listened_at DESC, user_id) WITH (timescaledb.transaction_per_chunk);
+CREATE UNIQUE INDEX listened_at_track_name_user_id_ndx_listen ON listen (listened_at DESC, track_name, user_id) WITH (timescaledb.transaction_per_chunk);
```
Comment (amCap1712) [inline admin/timescale/updates/2022-01-13-add-user-id-indexes.sql:7]: Only adding new indexes for now, existing ones will drop later when we remove user_name table.

File: listenbrainz/tests/integration/test_timescale_writer.py [hunk 2]
```diff
@@ -17,10 +18,11 @@
 from datetime import datetime
 
 
-class TimescaleWriterTestCase(IntegrationTestCase):
+class TimescaleWriterTestCase(IntegrationTestCase, TimescaleTestCase):
 
     def setUp(self):
-        super(TimescaleWriterTestCase, self).setUp()
+        IntegrationTestCase.setUp(self)
+        TimescaleTestCase.setUp(self)
```
Comment (amCap1712) [inline listenbrainz/tests/integration/test_timescale_writer.py:25]: Since now we are using user_ids instead of user names and user ids are `SERIAL`, we  get conflict in various tests. We disucssed in IRC and came up with the solution to make a util method which assigns random user ids and use it everywhere. Since, that is a wholesale change to make everywhere I'd prefer to do it in a separate PR as part of general improvements to tests.

Comment (mayhem) [reply listenbrainz/tests/integration/test_timescale_writer.py:25]: :+1:

**Review by mayhem — APPROVED** [review]
Slogging through it all!

**Comment by alastair** [comment]
Once all rows in the listens table have a user_id value set, we should remove the `default 0`. I didn't add this to create_tables.sql, only the migration script because it was a way to set up the column to be not null, We should remove the default to enforce that we always provide a value. not sure if this needs to go in part 2 or part 3

**Review by mayhem — APPROVED** [review]
All changed bits since last review are fine.

File: listenbrainz/webserver/views/user.py [hunk 3]
```diff
@@ -424,7 +424,7 @@ def delete_user(musicbrainz_id):
     """
 
     user = _get_user(musicbrainz_id)
-    timescale_connection._ts.delete(user.musicbrainz_id)
+    timescale_connection._ts.delete(musicbrainz_id, user.id)
```
Comment (alastair) [inline listenbrainz/webserver/views/user.py:427]: we should still use `user. musicbrainz_id` here, in case there are upper/lower-case differences between the username and what's in redis

**Review by alastair — COMMENTED** [review]
This is good, thanks for working through all of the tiny 1 line changes to make this happen.

One thing that I'm unsure about is that we're still using `listen.user_name` in a few places - in redis for listen counts, and when getting data from the listen table to turn in to Listen object (e.g. `fetch_listens_for_multiple_users_from_storage`).
This means that we're still going to have to make a further change to stop using this column before we delete it. It also means that when renaming users we still have to update the listen table so that these get methods return the correct data.

I think it'd be a good idea to make this pull request completely remove us from the dependency of the user_name field so that the only remaining thing that we need to do is remove the column.

File: listenbrainz/listenstore/timescale_utils.py [hunk 4]
```diff
@@ -60,31 +60,31 @@ def recalculate_all_user_data():
                 len(user_list))
 
     # Reset the timestamps and listen counts to 0 for all users
-    for user_name in user_list:
+    for user_name, user_id in user_list:
         cache.set(REDIS_USER_LISTEN_COUNT + user_name, 0, expirein=0, encode=False)
-        cache.set(REDIS_USER_LISTEN_COUNT + user_name, 0, expirein=0, encode=False)
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_utils.py:65]: Yeah, indeed.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 5]
```diff
@@ -987,20 +984,21 @@ def import_listens_dump(self, archive_path: str, threads: int = DUMP_DEFAULT_THR
 
         return total_imported
 
-    def delete(self, musicbrainz_id):
-        """ Delete all listens for user with specified MusicBrainz ID.
+    def delete(self, musicbrainz_id, user_id):
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:987]: Yes, right. Will update.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 6]
```diff
@@ -75,12 +75,12 @@ def __init__(self, conf, logger):
...
             'LISTEN_DUMP_TEMP_DIR_ROOT', tempfile.mkdtemp())
 
-    def set_empty_cache_values_for_user(self, user_name):
+    def set_empty_cache_values_for_user(self, user_name, user_id):
         """When a user is created, set the listen_count and timestamp keys so that we
            can avoid the expensive lookup for a brand new user."""
 
         cache.set(REDIS_USER_LISTEN_COUNT + user_name, 0, expirein=0, encode=False)
-        cache.set(REDIS_USER_TIMESTAMPS + user_name, "0,0", expirein=0)
+        cache.set(REDIS_USER_TIMESTAMPS + str(user_id), "0,0", expirein=0)
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:83]: Yes, this has changed in #1700.

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 7]
```diff
@@ -283,27 +280,27 @@ def insert(self, listens):
         # update the listen counts and timestamps for the users
         user_timestamps = {}
         user_counts = defaultdict(int)
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:282]: user_counts has been removed in #1700

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 8]
```diff
@@ -426,9 +423,9 @@ def fetch_listens_for_multiple_users_from_storage(self, user_names: List[str], f
             listens.reverse()
 
         self.log.info("fetch listens %s %.2fs (%d passes)" %
-                      (str(user_names), fetch_listens_time, passes))
+                      (str(user_ids), fetch_listens_time, passes))
 
-        return (listens, min_user_ts, max_user_ts)
+        return listens, min_user_ts, max_user_ts
 
     def fetch_recent_listens_for_users(self, user_list, limit=2, max_age=3600):
```
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:430]: I intend to do a separate PR to remove user_name bits and can remove user_name from here at that point? Also, note that we always get user_names in API and convert those to IDs so we have already performed the db query before reaching this method.

File: listenbrainz/webserver/views/user.py (hunk 3 above)
Comment (amCap1712) [reply listenbrainz/webserver/views/user.py:427]: Something weird is going on here, this function receives the musicbrainz_id not from what the user submitted but from a db query and then we do a db query here again. Will look deeper later for now changing user.musicbrainz_id

Comment (alastair) [reply listenbrainz/webserver/views/user.py:427]: yeah, I wouldn't be surprised if we go username -> look up user -> call this method -> lookup again. 
Making a ticket to pass in a User object here sounds good

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:987]: I checked again and there are 2 uses of delete. One when the user deletes themselves, then we have current_user. Second is when we admin delete. Further, this method is changing in #1700 to only accept user ids.

File: listenbrainz/webserver/views/user.py (hunk 3 above)
Comment (amCap1712) [reply listenbrainz/webserver/views/user.py]: LB-1055

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:987]: Added to LB-1055.

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alastair) [reply listenbrainz/listenstore/timescale_listenstore.py:430]: It's probably OK if we have a separate PR for removing user_name, although I think then we might need to do 2 more? 1 which stops reading listen.user_name (what I'm suggesting here), and then a final one to actually delete the column?

File: admin/timescale/updates/2022-01-13-add-user-id-indexes.sql (hunk 1 above)
Comment (amCap1712) [reply admin/timescale/updates/2022-01-13-add-user-id-indexes.sql]: Updated.

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:430]: Thinking more, deleting in separate PR makes sense. So yes, 2 PRs then. I am fine with not merging this till we finalize this change in another PR and then deploy it all together.

[the remark this answers is not in the document]
Comment (alastair) [reply listenbrainz/listenstore/timescale_listenstore.py:430]: https://github.com/metabrainz/listenbrainz-server/pull/1833
